### ISSUE TITLE
Is there a way to receive data from a job as it is output without polling?

### ISSUE THREAD

**Issue by alx9r** [PowerShell/PowerShell#6540]
The following code starts a job that outputs an integer once a second.  Subscriptions for some of the job's events are established.  The output of the job is polled by calling `Receive-Job` periodically until the job stops running.  The events are output.

```PowerShell
$j = Start-Job {1..5 | % {$_; sleep 1} }

Register-ObjectEvent -InputObject $j.Output -EventName 'DataAdded'
Register-ObjectEvent -InputObject $j -EventName StateChanged

while ( $j.State -eq [System.Management.Automation.JobState]::Running )
{
   $j | Receive-Job
   sleep 0.2
}

Get-Event | % { $_ | Remove-Event; $_ }
```

When the above code is run, an integer is output once per second followed by a single `StateChanged` event as follows:

```none
1
2
3
4
5

ComputerName     :
RunspaceId       : ef1f4543-86d6-410c-9d95-3883faeceacb
EventIdentifier  : 1
Sender           : System.Management.Automation.PSRemotingJob
SourceEventArgs  : System.Management.Automation.JobStateEventArgs
SourceArgs       : {Job1, System.Management.Automation.JobStateEventArgs}
SourceIdentifier : 3ece9623-2ef4-4f86-a7ce-bf088f483161
TimeGenerated    : 2018-04-01 5:59:56 PM
MessageData      :
```

This demonstrates the following:

* a `StateChanged` event is not published when new data is available to be received
* no `Job.Output.DataAdded` event is published despite that data was output by the job then received

In other words, neither of these two obvious APIs facilitate receiving data from a job as it is output.  I have also noticed that the `.Output` member of jobs seems to always be empty.

This leaves me with the following questions:

1. Is a way of subscribing to or waiting on new data that is added to a job without polling?
2. Why is no `DataAdded` event published when the above code is run?  In what conditions is that event published?
3. Why is the `.Output` property of a job object always empty?  In what conditions is that property populated?

**Comment by PetSerAl** [comment]
```powershell
$j = Start-Job {1..5 | % {$_; sleep 1} }

Register-ObjectEvent -InputObject $j.ChildJobs[0].Output -EventName 'DataAdded'
Register-ObjectEvent -InputObject $j -EventName StateChanged

$j | Receive-Job -Wait

Get-Event | % { $_ | Remove-Event; $_ }
```

**Comment by alx9r** [comment]
Thanks @PetSerAl.

For future reference, [about_Jobs_Details](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_job_details?view=powershell-6) includings the following statement (which I missed on first read):

> In jobs started by using Start-Job or the AsJob parameter of Invoke-Command, the parent job is an executive. It does not run any commands or return any results. The commands are actually run by the child jobs. (Jobs started by using other cmdlets might work differently.)

Subscribing to the `DataAdded` event on the child job as suggested by PetSerAl indeed works to receive data as it is output:

```PowerShell
$j = Start-Job {1..5 | % {$_; sleep 1} }

Register-ObjectEvent -InputObject $j.ChildJobs[0].Output -EventName 'DataAdded' -SourceIdentifier DataAdded
Register-ObjectEvent -InputObject $j -EventName StateChanged

while ( $j.State -eq [System.Management.Automation.JobState]::Running )
{
    Wait-Event |
        ? { $_.SourceIdentifier -eq 'DataAdded' } |
        % {
            $j | Receive-Job
            $_ | Remove-Event
        }
}

Get-Event | Remove-Event
Unregister-Event DataAdded
```

**Comment by DennisL68** [comment]
> ```powershell
> $j = Start-Job {1..5 | % {$_; sleep 1} }
> 
> Register-ObjectEvent -InputObject $j.ChildJobs[0].Output -EventName 'DataAdded' -SourceIdentifier DataAdded
> Register-ObjectEvent -InputObject $j -EventName StateChanged
> 
> while ( $j.State -eq [System.Management.Automation.JobState]::Running )
> {
>     Wait-Event |
>         ? { $_.SourceIdentifier -eq 'DataAdded' } |
>         % {
>             $j | Receive-Job
>             $_ | Remove-Event
>         }
> }
> 
> Get-Event | Remove-Event
> Unregister-Event DataAdded
> ```

This even works with remote jobs :)

Just tried using creating a job using Invoke-Command instead...

~~~
$j = Invoke-Command -ComputerName 'RemoteServer' -Scriptblock {1..5 | foreach {$_; sleep 1} } -AsJob
~~~~
